Fix generating password

speedlog opened the pull request:
Fixes #8

speedlog committed:
Fix generating password

cardil commented:
Please provide a test with random seeded with constant value

cardil reviewed (changes requested):
Please wait to #10 be merged, and then merge develop branch.

speedlog committed:
Merge remote-tracking branch 'origin/develop' into bugfix/fix-password-generator

speedlog committed:
Refactor of generating credentials

1) Remove using substring of uuid as username (substring may be not unique)
2) Provide random generator to UsernameAndPasswordCredentialsGeneratorService
3) Refactor - change store password from String to char[]

cardil commented on cascades-server/configuration/src/main/java/pl/gov/coi/cascades/server/presentation/PresentationConfiguration.java:
```diff
@@ -33,8 +36,13 @@ DatabaseNameGeneratorService produceDatabaseNameGeneratorService() {
+    }
+
+    @Bean
+    Random produceRandomGenerator() {
```
`Random` producer may be not needed as Spring Boot does this automatically, at least I think (need to verify).

If really needed it should be moved to some other Configuration class, more generic.

cardil commented on cascades-server/configuration/src/test/java/pl/gov/coi/cascades/server/presentation/PresentationConfigurationTest.java:
```diff
@@ -26,11 +30,10 @@
-    public void testProduceTemplateIdGeneratorService() throws Exception {
-        // given
-        PresentationConfiguration presentationConfiguration = new PresentationConfiguration();
+    private final PresentationConfiguration presentationConfiguration = new PresentationConfiguration();
```
In tests produce an instance by passing `Random` initialized with a constant seed, like that `new Random(123456L)`.

In that way you will be able to assert that produced password is exactly something like that: 'QAss23dd2sss'. It will not change until generation algorithm has, with you like to check.

cardil commented on cascades-server/configuration/src/test/java/pl/gov/coi/cascades/server/presentation/PresentationConfigurationTest.java:
```diff
@@ -66,23 +63,27 @@ public void testProduceDatabaseNameGeneratorService() throws Exception {
+    public void testProduceCredentials() {
         // when
-        UsernameAndPasswordCredentialsGeneratorService actual = presentationConfiguration.produceCredentials();
+        UsernameAndPasswordCredentialsGeneratorService actual = presentationConfiguration.produceCredentials(new SecureRandom());
```
Do not use `SecureRandom` in tests as it gives you no advantage and makes your tests run slower. Can be even, dramatically slower if you run out of entropy.

cardil commented on cascades-server/configuration/src/test/java/pl/gov/coi/cascades/server/presentation/PresentationConfigurationTest.java:
```diff
@@ -66,23 +63,27 @@ public void testProduceDatabaseNameGeneratorService() throws Exception {
+
+        // then
+        assertThat(actual).isNotNull();
+        assertThat(actual).isInstanceOf(ThreadLocalRandom.class);
```
There should be tests to assert for implementation, only for contract. In this case: `Random`. `ThreadLocalRandom` should not be concern here.

cardil commented on cascades-server/domain/logic/src/main/java/pl/gov/coi/cascades/server/domain/launchdatabase/UsernameAndPasswordCredentialsGeneratorService.java:
```diff
@@ -31,22 +31,21 @@ public UsernameAndPasswordCredentialsGeneratorService() {
     private static String generateUsername() {
-        String uuid = UUID.randomUUID().toString();
-        return uuid.substring(0, USERNAME_LENGTH);
+        return RandomStringUtils.randomAlphabetic(USERNAME_LENGTH);
```
`RandomStringUtils` should not be used as you don't control `Random` that it uses.

cardil commented on pom.xml:
```diff
@@ -154,6 +154,11 @@
                 <artifactId>org.apache.felix.framework</artifactId>
                 <version>5.6.2</version>
             </dependency>
+            <dependency>
```
Remove `commons-lang` as it is only used for one method that easily can be rewritten to be better. Also `commons-lang` is code from 2011. Bleh. :mask:
